### ISSUE TITLE
latency issues inside pod (in GCE)

### ISSUE THREAD

**Issue by JeanMertz** [kubernetes/kubernetes#14315]
We're running a Kubernetes cluster on top of GCE, and notice a very high latency of TCP packets. We tested this in multiple (Debian-based) running containers, each consistently giving these kind of results.

I am unsure if this is Kubernetes related, or Google Container Engine related, but this is causing us major headaches right now, any thoughts?

```
$ kubectl exec debian-tailf-xk8jy-mxl9t -it -- bash
$ cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 8 (jessie
```

```
$ curl -w "@format.txt" -o /dev/null -s "https://www.google.nl/"            
    time_namelookup:  3.523
       time_connect:  3.523
    time_appconnect:  3.555
   time_pretransfer:  3.555
      time_redirect:  0.000
 time_starttransfer:  3.580
                    ----------
         time_total:  3.581

$ curl -w "@format.txt" -o /dev/null -s "https://www.google.nl/"            
    time_namelookup:  8.554
       time_connect:  8.555
    time_appconnect:  8.560
   time_pretransfer:  8.560
      time_redirect:  0.000
 time_starttransfer:  8.581
                    ----------
         time_total:  8.581
```

```
curl -w "@format.txt" -o /dev/null -s  https://jquery-speedtest.googlecode.com/files/100MB.txt
    time_namelookup:  8.527
       time_connect:  8.533
    time_appconnect:  8.538
   time_pretransfer:  8.538
      time_redirect:  0.000
 time_starttransfer:  8.793
                    ----------
         time_total:  15.189
```

```
traceroute google.nl
traceroute to google.nl (64.233.166.94), 30 hops max, 60 byte packets
 1  10.92.0.1 (10.92.0.1)  0.029 ms  0.029 ms  0.021 ms
 2  wm-in-f94.1e100.net (64.233.166.94)  0.583 ms  0.536 ms  0.515 ms
```

From what I can tell, there are enough resources, and the cluster itself (three `n1-standard-4` nodes) have ~6 non-heavy pods running on top of them.

**Comment by thockin** [comment]
It looks like all the time is going into DNS?  @artfulcoder  We've had a
few such reports and no easy repro.

**Comment by JeanMertz** [comment]
I can confirm that this *only* happens inside pods, not on the GCE nodes itself:

```
curl -w "@format.txt" -o /dev/null -s "https://www.google.nl/"
    time_namelookup:  0.003
       time_connect:  0.004
    time_appconnect:  0.164
   time_pretransfer:  0.164
      time_redirect:  0.000
 time_starttransfer:  0.196
                    ----------
         time_total:  0.196
```

**Comment by JeanMertz** [comment]
Here's the top three processes on one of the nodes. I find the CPU consumption relatively high (for a relatively empty cluster of three `n1-standard-4` nodes), but perhaps this is as expected and unrelated

```
PID USER      PR  NI  VIRT  RES  SHR S  %CPU %MEM    TIME+  COMMAND
3411 root      20   0  739m 549m  20m S  69.6  3.6   4363:00 /usr/local/bin/kubelet --api_servers=https://xxx --enable-debugging-handlers=true --cloud_provider=gce --config=/etc/kubernetes/manifests --allow_privileged=False --v=2 --cluster_dns=10.95.240.10 --cluster_domain=cluster.local --configure-cbr0=true --cgroup_root=/ --system-container=/system
3130 root      20   0 1146m 880m  11m S  40.6  5.8 673:31.34 /usr/local/bin/kube-proxy --master=https://xxx --kubeconfig=/var/lib/kube-proxy/kubeconfig --v=2
3683 root      20   0 4081m 437m  12m S  34.8  2.9   1720:34 /usr/bin/docker -d -p /var/run/docker.pid --insecure-registry 10.0.0.0/8 --bridge=cbr0 --iptables=false --ip-masq=false
```

**Comment by thockin** [comment]
That's REALLY high CPU usage.  What's going on?  Can you look at the logs
for kubelet and / or kube-proxy?

**Comment by JeanMertz** [comment]
For `kube-proxy`, this might be related to the original cause of this ticket:

```
I0922 05:31:53.732978    3130 roundrobin.go:262] LoadBalancerRR: Setting endpoints for production/api:http to [10.92.0.224:80 10.92.1.227:80 10.92.2.178:80 10.92.2.179:80 10.92.2.181:80]
I0922 05:31:53.733191    3130 roundrobin.go:219] Delete endpoint 10.92.1.227:80 for service "production/api:http"
I0922 05:31:59.642296    3130 proxysocket.go:246] New UDP connection from 10.92.0.234:44428

last line x 70, with different ports
```

One of our RC's has a `readinessProbe` attached to its pods, but due to the high latency, every ~10 seconds a pod gets removed (and later added back) from the LB. The above log appears instantly every ~10 seconds for different IPs.

We already increased the timeout period to 5, but it doesn't help.

**Comment by thockin** [comment]
Those are just UDP connections being made through the proxy.  It shouldn't
be a big issue.  That said, we have received a lot of similar bug reports -
DNS taking multiple seconds.  But I can not repro it...

**Comment by JeanMertz** [comment]
With the probe disabled, I still see these loads (fluctuating highly):

```
 3411 root      20   0  739m 550m  20m S 114.7  3.6   4386:55 /usr/local/bin/kubelet --api_servers=https://xxx --enable-debugging-handlers=true --cloud_provider=gce --config=/etc/kubernetes/manifests --allow_privileged=False --v=2 --cluster_dns=10.95.240.10 --cluster_domain=cluster.local --configure-cbr0=true --cgroup_root=/ --system-container=/system
 3683 root      20   0 4085m 441m  12m S 102.0  2.9   1727:46 /usr/bin/docker -d -p /var/run/docker.pid --insecure-registry 10.0.0.0/8 --bridge=cbr0 --iptables=false --ip-masq=false
 6253 root      20   0  186m 172m  10m R  63.7  1.1   0:13.59 /heapster --source=kubernetes:''
 3130 root      20   0 1165m 895m  11m S  51.0  5.9 689:38.38 /usr/local/bin/kube-proxy --master=https://xxx --kubeconfig=/var/lib/kube-proxy/kubeconfig --v=2
```

Some more observations:

* using `apt-get update` on a pod, I can see it hanging a long time on `Connecting to httpredir.debian.org` (again confirming initial dns issue)
* looking at the k8s web ui, I can see the CPU chart highly fluctuating between 30% and 100% on all three nodes, memory is consistently at ~75%
* on the same web ui, the `events` tab is empty, `services` only shows "kubernetes", `nodes` shows the correct nodes, `pods` is empty
* here are the details of one node as shown in the web ui:

        Capacity	
        cpu:	4
        memory:	15437216Ki
        pods:	40
        System Info	
          containerRuntimeVersion:	docker://Unknown
          kernelVersion:	3.16.0-0.bpo.4-amd64
          kubeProxyVersion:	v1.0.4
          kubeletVersion:	v1.0.4
        machineID:	
        osImage:	Debian GNU/Linux 7 (wheezy)

* showing the dns issue right now:

        $ time ping -c3 google.nl

        PING google.nl (74.125.133.94): 56 data bytes
        64 bytes from 74.125.133.94: icmp_seq=0 ttl=54 time=7.505 ms
        64 bytes from 74.125.133.94: icmp_seq=1 ttl=54 time=0.526 ms
        64 bytes from 74.125.133.94: icmp_seq=2 ttl=54 time=0.589 ms
        --- google.nl ping statistics ---
        3 packets transmitted, 3 packets received, 0% packet loss
        round-trip min/avg/max/stddev = 0.526/2.873/7.505/3.275 ms

        real	0m37.061s
        user	0m0.000s
        sys	0m0.000s


I'm open to a pair-debuging session on our setup, if that helps?

**Comment by thockin** [comment]
I am out of office tomorrow, but I wouldn't mind poking around later in the
week.  If nobody has better ideas, of course.

**Comment by JeanMertz** [comment]
@thockin that would be great, as I have yet to solve this.

I've got a Kubernetes cluster in which networking is truely acting weird, it's set up from scratch only a couple of days ago, using Google Container Engine, without any manual tweaking.

**Comment by JeanMertz** [comment]
We just removed one of the nodes that had networking issues, things seem to work better now, but I'll continue to review what happens.

This was a GKE-provided node, without any manual modifications.

**Comment by thockin** [comment]
If you can provide more details maybe we can look into what's going on.  You can email me (my GH name @google.com) with your project name and ... what else @roberthbailey ?

**Comment by roberthbailey** [comment]
Project name is sufficient if you only have one cluster. If you have more than one, then zone and/or cluster name (enough that it makes this cluster unique within your project).

**Comment by ArtfulCoder** [comment]
@JeanMertz 

Per my investigations on my own GCE cluster, the DNS issues seem to happen because : 
*  All DNS queries from kuberenetes pod, go through a skydns pod that Kubernetes runs.
SkyDNS forwards queries to the google nameserver which throttles requests sometimes because it  sees too many requests coming from the same IP (skydns pod ip)

* Skydns times out on throttled requests and then retries. Most times, the retry works and the dns client gets a response.

* There was a bug in skydns that caused the timeout to be 4 seconds (the intent was to timeout in 2 seconds). In release 1.1, we have addressed this issue. #14625

Post v1.1, we will pick a new skydns release where skydns will cached responses on forwarded requests. This should mitigate the throttling issue.

Closing this bug. 
Fix: upgrade to v1.1 (when released)
